### PR TITLE
core, crypto, params: implement CREATE2 evm instrction

### PR DESCRIPTION
**Pull request by rjl493456442**
Implements [EIP1014](https://github.com/ethereum/EIPs/pull/1014)

### CODE REVIEW

File: core/vm/opcodes.go [hunk 1]
```diff
@@ -209,6 +209,7 @@ const (
 	CALLCODE
 	RETURN
 	DELEGATECALL
+	CREATE2
```
Comment (karalabe) [inline core/vm/opcodes.go]: Please skim through this file, there are 1 or 2 more instances where you need to add the opcode (the string to opcode and opcode to string mappings).

**Comment by karalabe** [comment]
Small issues, otherwise LGTM

**Comment by holiman** [comment]
Within op Create2, no need to check for eip150 or homestead. There is no create2 prior to Constantinople.

**Comment by karalabe** [comment]
I did think about that, but I don't think we have code that auto-enables
previous forks if you enable a specific one. We should implement it though
imho.

**Comment by holiman** [comment]
Can you use a fixed size buffer for salt instead of slice? It's defined by the eip, right?

**Comment by rjl493456442** [comment]
@holiman Sorry i didn't find the salt definition in the EIP

File: crypto/crypto.go [hunk 2]
```diff
@@ -76,6 +76,14 @@ func CreateAddress(b common.Address, nonce uint64) common.Address {
 	return common.BytesToAddress(Keccak256(data)[12:])
 }
 
+// CreateAddress2 creates an ethereum address given the address bytes, initial
+// contract code and a salt.
+// TODO(rjl493456442) considering address collision, should we add the 0xff as the input prefix?
+func CreateAddress2(b common.Address, salt []byte, code []byte) common.Address {
+	data, _ := rlp.EncodeToBytes([]interface{}{b, salt, code})
+	return common.BytesToAddress(Keccak256(data)[12:])
```
Comment (karalabe) [inline crypto/crypto.go]: I'm unsure whether this method is correct or not. The spec seems to state `sha3(msg.sender ++ salt ++ init_code)[12:]`. Given that `sender`, `salt` and `code` are all bytes, I'd interpret the formula as simple append, not RLP encoding. Might be worth a clarification on the EIP.

Comment (rjl493456442) [reply crypto/crypto.go]: For the rlp encoding, no matter address is `common.Address` or byte slice, the result should be same.
And RLP or simple Append, yes we should give some feedbacks to EIP and make a clarification

Comment (karalabe) [reply crypto/crypto.go]: https://github.com/ethereum/EIPs/pull/1014#issuecomment-406193182

File: core/vm/instructions.go [hunk 3]
```diff
@@ -665,6 +665,40 @@ func opCreate(pc *uint64, evm *EVM, contract *Contract, memory *Memory, stack *S
...
 
+func opCreate2(pc *uint64, evm *EVM, contract *Contract, memory *Memory, stack *Stack) ([]byte, error) {
+	var (
+		endowment    = stack.pop()
+		offset, size = stack.pop(), stack.pop()
+		salt         = stack.pop()
+		input        = memory.Get(offset.Int64(), size.Int64())
+		gas          = contract.Gas
+	)
+	if evm.ChainConfig().IsEIP150(evm.BlockNumber) {
```
Comment (holiman) [inline core/vm/instructions.go]: This is 'speculative' -- how would Create2 work before EIP150? That's not defined, so I'd vote to remove this.

File: core/vm/instructions.go [hunk 4]
```diff
@@ -665,6 +665,40 @@ func opCreate(pc *uint64, evm *EVM, contract *Contract, memory *Memory, stack *S
...
+		gas -= gas / 64
+	}
+
+	contract.UseGas(gas)
+	res, addr, returnGas, suberr := evm.Create2(contract, input, gas, endowment, salt)
+	// Push item on the stack based on the returned error. If the ruleset is
+	// homestead we must check for CodeStoreOutOfGasError (homestead only
+	// rule) and treat as an error, if the ruleset is frontier we must
+	// ignore this error and pretend the operation was successful.
+	if evm.ChainConfig().IsHomestead(evm.BlockNumber) && suberr == ErrCodeStoreOutOfGas {
```
Comment (holiman) [inline core/vm/instructions.go]: Same as above. Yes, it's (post) homestead

**Comment by rjl493456442** [comment]
@holiman @karalabe Updated, PTAL again.

**Comment by karalabe** [comment]
We'll need to extend the call tracer to be able to handle this new create. But lets do that in a followup pr.

**Comment by rjl493456442** [comment]
@karalabe Will do!

**Comment by holiman** [comment]
Lgtm!

**Pull request merged**